Adds Professor DUMMY to the CMO office, accessible by the CMO and the SEA

Vicacrov opened the pull request:
# About the pull request

Adds a new wall cabinet to the CMO's office called the "professor dummy cabinet". It contains Professor DUMMY and its tablet: it is a tool for new nurses(!!) and doctors to learn medical treatment, including surgery.

The locker can only be opened by the CMO or the SEA (not even the CO can open it). It cannot be flashbanged, melted, or slashed.

There is a hefty warning for everyone when the dummy falls out, in case someone destroys the closet in a creative way:

![image](https://github.com/cmss13-devs/cmss13/assets/49321394/632c6bff-02dc-428b-9d21-4f5acedd9db8)

The dummy and its tablet dust upon hijack impact, to avoid either side having a godmoded mob.

# Explain why it's good for the game

It is a tool for new medical personnel to practice on. It can also ease boredom if the round is going too smoothly for marines.

# Testing Photographs and Procedure

Setting up the tablet and hijacking. (The dummy no longer spawns in the Morgue but in the CMO office).

<details>
<summary>Screenshots & Videos</summary>

https://github.com/cmss13-devs/cmss13/assets/49321394/563db21d-ebc2-41e0-a8e8-764f9b8aa61b

</details>


# Changelog

:cl:
add: Added a professor dummy cabinet to the CMO office, containing Professor DUMMY and its control tablet. It is a tool to teach new medical personnel with. The cabinet is accessible by the CMO and the SEA only.
/:cl:

ihatethisengine commented:
maybe just make it uninteractable on hijack?

Vicacrov commented:
> maybe just make it uninteractable on hijack?

Dusting might be a bit jarring, but having a human-looking mob nobody can interact with would be more disruptive, I think.

Drulikar reviewed (changes requested):
This needs to instead be in the SEA vendor where its unhackable and otherwise then falls under the responsibility of the SEA that lent it to someone to use to put it away.

If you also want CMOs to access it, then they'd need a similar situation where only they (or a SEA) can vend it.

As of now, its the responsibility of the admin that spawned it to ensure it doesn't get misused.

Vicacrov commented:
Implemented changes:
- Made a new subtype for this dummy as I realised the character setup one was the same.
- Reverted the Morgue changes.
- The dummy now dusts upon hijack impact, not upon hijack start (code bloat on my part).
- Added a special closet to the CMO office, the "professor dummy cabinet".
  - It can be only opened with a CMO or SEA **job**, not access. (Not even a CO ID helps here.)
  - It cannot be flashbanged, slashed, or melted.
  - When the dummy falls out, everyone around the closet gets a warning how it should not be misused - this also covers the edge case where it gets destroyed by unconventional means. Ignorance is not an excuse when you get a red text this big.
  - The closet can store mobs but it fails to close if there is any other mob on its tile but the dummy. (No hiding in the indestructible closet.)

![image](https://github.com/cmss13-devs/cmss13/assets/49321394/395313d0-9c95-4a7c-a75e-585e4c814223)

![image](https://github.com/cmss13-devs/cmss13/assets/49321394/632c6bff-02dc-428b-9d21-4f5acedd9db8)

Nanu308 commented on maps/map_files/USS_Almayer/USS_Almayer.dmm:
mapping lgtm

Drulikar commented on code/modules/mob/living/carbon/human/human_dummy.dm:
```diff
@@ -74,11 +74,42 @@ GLOBAL_LIST_EMPTY(dummy_mob_list)
+	regenerate_icons()
+
+	// Ensure nurses can practice surgery on it
+	mob_flags |= EASY_SURGERY
```
You are missing `huggable = FALSE` like `/datum/equipment_preset/other/professor_dummy/load_race` used to do.

Drulikar commented on code/game/objects/structures/crates_lockers/closets/secure/medical.dm:
```diff
@@ -171,3 +171,54 @@
+/obj/structure/closet/secure_closet/professor_dummy/flashbang(datum/source, obj/item/explosive/grenade/flashbang/FB)
+	return
+
+/obj/structure/closet/secure_closet/professor_dummy/emp_act(severity)
```
ex_act also needs to be implemented the same way. Throw a HE grenade near it and it'll dump the contents out.

Drulikar commented on code/modules/gear_presets/other.dm:
```diff
@@ -845,38 +845,6 @@
 
 //*****************************************************************************************************/
 
-/datum/equipment_preset/other/professor_dummy
```
Removing this means it is no longer in the create humans menu; which is generally how its been spawned. It'd probably be better to just keep the preset but have your closest spawn the equipment preset into the closet.

Drulikar replied:
See https://github.com/cmss13-devs/cmss13/blob/master/code/modules/admin/topic/topic_events.dm#L151-L159 how create_humans uses these presets. You likely just need to create the mob, and then `arm_equipment` it.

Drulikar replied:
If the concern was just not registering the signal for admin spawned; then have the closet register the signal for that spawn on the dummy's behalf (as in we aren't registering a signal on the closet, but the closet's initialize call is having the dummy register the signal)

Drulikar replied:
Relayed on discord, but putting it here too. Something like
```js
/obj/structure/closet/secure_closet/professor_dummy/Initialize()
    . = ..()
    var/mob/living/carbon/human/dummy/professor_dummy/spawned_dummy = new(src)
    var/datum/equipment_preset/dummy_preset = /datum/equipment_preset/other/professor_dummy
    arm_equipment(spawned_dummy, initial(dummy_preset.name), TRUE, FALSE)
    spawned_dummy.RegisterSignal(SSdcs, COMSIG_GLOB_HIJACK_LANDED, TYPE_PROC_REF(/mob/living/carbon/human/dummy/professor_dummy, destroy_upon_hijack))```
should work. The signal registration here is just if we want only closet dummies to dust; but I don't mind admin spawns having the same signal either if they're on a game level z.

Drulikar commented on code/game/objects/structures/crates_lockers/closets/secure/medical.dm:
```diff
@@ -171,3 +171,54 @@
+	..()
+
+/obj/structure/closet/secure_closet/professor_dummy/close()
+	for(var/mob/M in loc)
+		if(!istype(M, /mob/living/carbon/human/dummy/professor_dummy))
```
```suggestion
	for(var/mob/current_mob in loc)
		if(!istype(current_mob, /mob/living/carbon/human/dummy/professor_dummy))
```
No single letter vars please: https://github.com/cmss13-devs/cmss13/blob/master/.github/guides/STYLES.md#dont-use-abbreviations

Drulikar commented on code/game/objects/structures/crates_lockers/closets/secure/medical.dm:
```diff
@@ -171,3 +171,54 @@
+/obj/structure/closet/secure_closet/professor_dummy/dump_contents()
+	if(locate(/mob/living/carbon/human/dummy/professor_dummy) in src)
+		visible_message(SPAN_HIGHDANGER("Professor DUMMY should only be used for teaching medical personnel, exclusively done by the [JOB_CMO] or the [JOB_SEA]. Do not abuse it."))
+	..()
```
```suggestion
	return ..()
```
The other implementations are returning TRUE/FALSE/1/0

Drulikar commented on code/game/objects/structures/crates_lockers/closets/secure/medical.dm:
```diff
@@ -171,3 +171,54 @@
+
+/obj/structure/closet/secure_closet/professor_dummy/togglelock(mob/living/user)
+	if(user.job == JOB_CMO || user.job == JOB_SEA)
+		..()
+		return
```
```suggestion
		return ..()
```
Although nothing normally is returned, since you're returning anyways you may as well.

Drulikar commented on code/modules/mob/living/carbon/human/human_dummy.dm:
```diff
@@ -74,11 +74,42 @@ GLOBAL_LIST_EMPTY(dummy_mob_list)
+	var/obj/item/device/professor_dummy_tablet/tablet = new (loc)
+	tablet.link_dummy(src)
+
+	if(is_mainship_level(z))
```
A closet's loc is not considered is_mainship_level you need to get_turf

Drulikar commented on code/game/objects/structures/crates_lockers/closets/secure/medical.dm:
```diff
@@ -171,3 +171,54 @@
+	return
+
+/obj/structure/closet/secure_closet/professor_dummy/emp_act(severity)
+	for(var/mob/living/carbon/human/dummy/professor_dummy/dummy in src)
```
If there only will ever be one dummy we can instead do something like
```js
var/mob/dummy = locate(/mob/living/carbon/human/dummy/professor_dummy) in src
if(dummy)
	qdel(dummy)
```

Drulikar commented on code/game/objects/structures/crates_lockers/closets/secure/medical.dm:
```diff
@@ -171,3 +171,64 @@
+	var/mob/living/carbon/human/dummy/professor_dummy/spawned_dummy = new(src)
+	var/datum/equipment_preset/dummy_preset = /datum/equipment_preset/other/professor_dummy
+	arm_equipment(spawned_dummy, initial(dummy_preset.name), TRUE, FALSE)
+	spawned_dummy.RegisterSignal(SSdcs, COMSIG_GLOB_HIJACK_LANDED, TYPE_PROC_REF(/mob/living/carbon/human/dummy/professor_dummy, destroy_upon_hijack))
```
```suggestion
```

Drulikar commented on code/modules/mob/living/carbon/human/human_dummy.dm:
```diff
@@ -74,11 +74,21 @@ GLOBAL_LIST_EMPTY(dummy_mob_list)
 	anchored = TRUE
+
+// Professor Dummy, used by CMOs and SEAs to teach new nurses/doctors
+/mob/living/carbon/human/dummy/professor_dummy/proc/destroy_upon_hijack()
```
```suggestion
/mob/living/carbon/human/dummy/professor_dummy/Initialize(mapload)
	. = ..()
	RegisterSignal(SSdcs, COMSIG_GLOB_HIJACK_LANDED, PROC_REF(destroy_upon_hijack))

/mob/living/carbon/human/dummy/professor_dummy/proc/destroy_upon_hijack()
```

Drulikar commented:
Feel free to undraft (Ready for review button near the bottom) when you're happy with it.
